src/runtime_src/hip/core/memory.h

```diff
@@ -1,17 +1,229 @@
 // SPDX-License-Identifier: Apache-2.0
-// Copyright (C) 2023 Advanced Micro Device, Inc. All rights reserved.
+// Copyright (C) 2024 Advanced Micro Device, Inc. All rights reserved.
 #ifndef xrthip_memory_h
 #define xrthip_memory_h
 
+#include "core/common/device.h"
+#include "experimental/xrt_bo.h"
+#include "experimental/xrt_ext.h"
+#include "device.h"
+#include "xrt/config.h"
+#include "xrt/device/hal.h"
+#include "xrt/util/range.h"
 
-namespace xrt::core::hip {
+namespace xrt::core::hip
+{
+  enum class memory_type : int
+  {
+    hip_memory_type_host = 0,
+    hip_memory_type_device,
+    hip_memory_type_managed,
+    hip_memory_type_registered,
+    hip_memory_type_invalid
+  };
+
+  class memory
+  {
+
+  public:
+    memory(std::shared_ptr<xrt::core::hip::device> dev)
+      : m_size(0), m_type(memory_type::hip_memory_type_invalid), m_hip_flags(0), m_host_mem(nullptr), m_device(dev), m_bo(nullptr), m_sync_host_mem_required(false)
+    {
+      assert(m_device);
+      init_xrt_bo();
+    }
+
+    memory(size_t sz, unsigned int flags, std::shared_ptr<xrt::core::hip::device> dev);
+
+    // construct from user host buffer
+    memory(size_t sz, void *host_mem, unsigned int flags, std::shared_ptr<xrt::core::hip::device> dev)
```

Comment:
ditto, Please change the order of params

Reply:
fixed.